File: my-folder/0200-number-of-islands/solution.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        total_islands=0
        rows=len(grid)
        cols=len(grid[0])
        def dfs(r,c):
            if(min(r,c)<0 or r>=rows or c>=cols or grid[r][c]=="0"):
                return 
            #mark the visited node with 0 so you dont go there in next iteration
            grid[r][c]="0"
            dfs(r+1,c)
            dfs(r-1,c)
            dfs(r,c+1)
            dfs(r,c-1)
        for r in range(rows):
            for c in range(cols):
                if(grid[r][c]=="1"):
                    dfs(r,c)
                    total_islands+=1

        return total_islands

        

File: my-folder/0200-number-of-islands/solution.java
class Solution {
    public int numIslands(char[][] grid) {
      int islands = 0;
      for(int row = 0; row < grid.length; row++){
        for(int col = 0; col < grid[0].length; col++){
            if(grid[row][col] == '1'){
                islands++;
                dfs(grid, row, col);
            }
        }
      }


      return islands;  
    }

    private void dfs(char[][] grid, int row, int col){
        if(row >= grid.length || col >= grid[0].length || row < 0 || col < 0 || grid[row][col] != '1'){
            return;
        }
        grid[row][col] = '0';
        dfs(grid, row + 1, col);
        dfs(grid, row - 1, col);
        dfs(grid, row, col + 1);
        dfs(grid, row, col - 1);
    }
}

//   ["1","1","1","1","0"]
//   ["1","1","0","1","0"]
//   ["1","1","0","0","0"]
//   ["0","0","0","0","0"]
